### PR TITLE
fuse-waked: allow directories that appear to be empty to rmdir()

### PR DESCRIPTION
**Pull request by mmjconolly**
If a directory existed on the underlying filesystem and was non-empty, but appeared to be empty within wakebox/fuse-waked, then fuse-waked would disallow the rmdir() even though it appeared to be empty. This could confuse a process that would think it's entitled to remove a writeable empty dir it had created (note: this means the directory is not in the 'visible' list).

This PR attempts to allow the _fiction_ of a successful rmdir() by letting the fuse client believe that the directory is gone, even though the underlying storage is left as-is.

Example of the issue:
```
$ cat bug.json
{
  "command": [
    "bash",
    "-c",
    "set -x; mkdir testdir/foo; ls -l -a testdir/foo; rm -rf testdir/foo"
  ],
  "environment": [ "PATH=/usr/bin:/bin" ],
  "directory": ".",
  "mount-ops": [
     {
      "type": "workspace",
      "destination": "."
     }
  ],
  "visible": [
      "testdir"
  ]
}
$ mkdir -p testdir/foo
$ touch testdir/foo/baa
$ wakebox -p bug.json
+ mkdir testdir/foo
+ ls -l -a testdir/foo
total 8
drwxrwxr-x. 2 root root 4096 Jun 10 18:43 .
drwxrwxr-x. 3 root root 4096 Jun 10 18:43 ..
+ rm -rf testdir/foo
rm: cannot remove 'testdir/foo': Directory not empty
$ echo $?
1
$ rm testdir/foo/baa
$ wakebox -p bug.json 
+ mkdir testdir/foo
+ ls -l -a testdir/foo
total 8
drwxrwxr-x. 2 matthewc matthewc 4096 Jun 10 18:51 .
drwxrwxr-x. 3 matthewc matthewc 4096 Jun 10 18:51 ..
+ rm -rf testdir/foo
$ echo $?
0
```

### CODE REVIEW

File: tools/fuse-waked/fuse-waked.cpp [hunk 1]
```diff
@@ -697,6 +697,19 @@ static int wakefuse_unlink_trace(const char *path)
 	return out;
 }
 
+// Check if the directory has children that were written and not yet deleted.
+// Note: we don't check the 'files_visible' list as when the directory itself or
+// a file within that directory is visible then is_writable() would have already failed.
+static bool has_written_children(const std::string& dir, Job& job) {
```
Comment (richardxia) [inline tools/fuse-waked/fuse-waked.cpp:703]:
```suggestion
static bool has_written_children(const std::string& dir, const Job& job) {
```

I think this can be a const reference?

File: tools/fuse-waked/fuse-waked.cpp [hunk 2]
```diff
@@ -697,6 +697,19 @@ static int wakefuse_unlink_trace(const char *path)
...
+	for (auto &written : job.files_wrote) {
+		if (written.size() > dir.size() && written[dir.size()] == '/' && written.find(dir) == 0) {
```
Comment (richardxia) [inline tools/fuse-waked/fuse-waked.cpp:705]: `find()` could be slow because it'll attempt to continue searching past the 0th index, even though that you only care about whether `dir` is a proper prefix of `written`. It's too bad that only C++20 added [`starts_with()`](https://en.cppreference.com/w/cpp/string/basic_string/starts_with), but I think you can at least use [`compare()`](https://cplusplus.com/reference/string/string/compare/) and just specify the start position and length to compare.

**Review by richardxia — APPROVED** [review]
Thanks! I can confirm that this fixes the both the original bug in SiFive's internal codebase as well as the small test case I created to reproduce the issue.

Is there a way for us to add that as a test in Wake's CI tests? I think we could create an even simpler test case than the one I narrowed by just creating an empty directory prior to running a wake target that attempts to remove it.

**Comment by JakeSiFive** [comment]
If a process removes a file what happens right now? Does it remove the file from `files_wrote`?

Also don't we still need to error out if there was a visible file in that directory? As is a bad process that tried to delete a folder that was populated even from its perspective would be allowed to think that it succeeded in deleting that file and then it could later read the directory and succeed when it shouldn't have.

File: tools/fuse-waked/fuse-waked.cpp (hunk 2 above)
Comment (terpstra) [reply tools/fuse-waked/fuse-waked.cpp:705]: Not to mention the outer loop is unnecessary. Try `lower_bound(dir + "/")` followed by check vs. `end` and a `compare`.

**Review by JakeSiFive — APPROVED** [review]
Happy with the code, the fact that it works for Richard, and Wes's explanation. Lets land it.

**Review by JakeSiFive — APPROVED** [review]
